src/main/java/school/hei/haapi/service/AwardedCourseService.java

```diff
@@ -87,4 +89,44 @@ public List<AwardedCourse> getByCriteria(
         PageRequest.of(page.getValue() - 1, pageSize.getValue(), Sort.by(ASC, "name"));
     return awardedCourseDao.findByCriteria(teacherId, courseId, pageable);
   }
+
+  public List<AwardedCourse> getByTeacherId(
+      String teacherId, PageFromOne page, BoundedPageSize pageSize) {
+    Pageable pageable =
+        PageRequest.of(page.getValue() - 1, pageSize.getValue(), Sort.by(ASC, "course"));
+    return awardedCourseRepository.findByMainTeacherId(teacherId, pageable);
+  }
+
+  @Transactional
+  public List<AwardedCourse> createOrUpdateAwardedCoursesByTeacherId(
+      String teacherId, List<AwardedCourse> awardedCourses) {
+
+    List<AwardedCourse> updatedAwardedCourses =
+        awardedCourses.stream()
+            .map(
+                awardedCourse -> {
+                  if (awardedCourse.getId() != null) {
+                    AwardedCourse existingAwardedCourse = findById(awardedCourse.getId());
+                    existingAwardedCourse.setCourse(awardedCourse.getCourse());
+                    existingAwardedCourse.setGroup(awardedCourse.getGroup());
+                    existingAwardedCourse.setMainTeacher(awardedCourse.getMainTeacher());
+                    awardedCourseValidator.accept(existingAwardedCourse);
```

Comment:
Here, you check if the awardedCourse already existed not if the id is null or not.
see what is Optional in java

Reply:
awardedCourseValidator.accept(...) can take a list as a param